lib/vagrant-libvirt/driver.rb

```diff
@@ -125,6 +140,65 @@ def get_domain_ipaddress(machine, domain)
         ip_address
       end
 
+      def restore_snapshot(machine, snapshot_name)
+        domain = get_domain(machine)
+        snapshot = get_snapshot_if_exists(machine, snapshot_name)
+        begin
+          # 4 is VIR_DOMAIN_SNAPSHOT_REVERT_FORCE
+          # needed due to https://bugzilla.redhat.com/show_bug.cgi?id=1006886
+          domain.revert_to_snapshot(snapshot, 4)
+        rescue Fog::Errors::Error => e
+          raise Errors::SnapshotReversionError, error_message: e.message
+        end
+      end
+
+      def is_snapshot_mode_on?(machine)
+        get_snapshot_if_exists(machine)
+        return true
+      rescue Errors::SnapshotMissing
+        return false
+      end
+
+      def list_snapshots(machine)
+        get_domain(machine).list_snapshots
+      rescue Fog::Errors::Error => e
+        raise Errors::SnapshotListError, error_message: e.message
+      end
+
+      def delete_snapshot(machine, snapshot_name)
+        get_snapshot_if_exists(machine, snapshot_name).delete
+      rescue Fog::Errors::Error => e
+        raise Errors::SnapshotDeletionError, error_message: e.message
+      end
+
+      def create_new_snapshot(machine, snapshot_name)
+        snapshot_desc = <<-EOF
+        <domainsnapshot>
+          <name>#{snapshot_name}</name>
+          <description>Snapshot for vagrant sandbox</description>
+        </domainsnapshot>
+        EOF
+        get_domain(machine).snapshot_create_xml(snapshot_desc)
```

Comment:
I made a mistake, the domain returned above by `get_client_domain` is not the same domain as `get_domain`, in fact it's possible that function was badly named and should have been `get_server`, since it is the fog server object that is returned there.

So will need to switch back, however can you make sure to pass the `machine` variable to `get_client_domain` and make that function private for now until sure it should be exposed directly.